File: src/leetcode/java/BestTimeToBuyAndSellStock.java
package leetcode.java;

public class BestTimeToBuyAndSellStock {
    /**
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;

        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            } else if (prices[i] - minPrice > maxProfit) {
                maxProfit = prices[i] - minPrice;
            }
        }

        return maxProfit;
    }

    /**
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     */
    public static int maxProfitWithSlidingWindow(int[] prices) {
        int maxProfit = 0;
        int left = 0;
        int right = 1;

        while (right < prices.length) {
            if (prices[left] < prices[right]) {
                maxProfit = Math.max(maxProfit, prices[right] - prices[left]);
            } else {
                left = right;
            }
            right++;
        }

        return maxProfit;
    }
}

File: src/leetcode/kotlin/BestTimeToBuyAndSellStock.kt
package leetcode.kotlin

class BestTimeToBuyAndSellStock {
    /**
     * Time complexity: O(n)
     * Space complexity: O(1)
     */
    fun maxProfitWithSlidingWindow(prices: IntArray): Int {
        var minPrice = Int.MAX_VALUE
        var maxProfit = 0
        var right = 0
        while (right < prices.size) {
            if (prices[right] < minPrice) {
                minPrice = prices[right]
            } else if (maxProfit < prices[right] - minPrice) {
                maxProfit = prices[right] - minPrice
            }

            right++
        }

        return maxProfit
    }
}